### PR TITLE
fix(http): fix memory leak in REST API

### PR DESCRIPTION
**Pull request by glasstiger**
Worker threads pool HttpConnectionContext objects in separate pools but they can steal each other context objects through the disconnect queue which is shared.
It can happen that the context pool of a worker becomes eventually empty in which case it creates a new object to be able to handle the request. New object means memory allocation.
Context objects are closed only when QuestDB stops so the context pools potentially can grow really big.
By limiting the pool size and closing contexts we are avoiding memory leak.

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 6 / 6 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/WeakObjectPool.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cutlass/http/HttpServer.java|5|5|100.00%|

**Review by bluestreak01 — DISMISSED** [review]

File: core/src/main/java/io/questdb/cutlass/http/HttpServer.java [hunk 1]
```diff
@@ -392,7 +395,7 @@ public HttpConnectionContext newInstance(long fd, IODispatcher<HttpConnectionCon
 
         @Override
         public void done(HttpConnectionContext context) {
-            if (closed) {
+            if (closed || contextPool.get().currentSize() > connectionPoolMaxCapacity) {
```
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/cutlass/http/HttpServer.java:398]: nit: we have the same potential problem in other places where we use `WeakObjectPool` and `WeakAutoClosableObjectPool`, e.g. `LineTcpConnectionContextFactory#LineTcpReceiver`. Does it make sense to store the max size in the pool classes themselves and modify `push()` method, so it gets `offer()` behavior?

Comment (glasstiger) [reply core/src/main/java/io/questdb/cutlass/http/HttpServer.java]: moved max size check into the pool

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 29 / 29 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/WeakObjectPool.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/cutlass/pgwire/AbstractTypeContainer.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/std/WeakObjectPoolBase.java|22|22|100.00%|
|:large_blue_circle:|io/questdb/std/WeakAutoClosableObjectPool.java|2|2|100.00%|

File: core/src/main/java/io/questdb/std/WeakObjectPoolBase.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+ * weak object pool is a factory of new instances, when object is popped
+ * out of pool there is no reference to tie this object back to pool.
+ * <p>
+ * In multithreaded environment weak pools must be local to thread and
+ * objects they create can travel between pools, e.g. one thread gets
+ * object from pool and last thread to finish with this object then
+ * returns it to its own local pool.
+ * <p>
+ * In an extreme situation this object migration could keep feeding some
+ * pools until we run out of memory. Hence we maximize the number of
```
Comment (puzpuzpuz) [inline core/src/main/java/io/questdb/std/WeakObjectPoolBase.java:41]: nit: "limit the maximum number" may sound more straightforward that "maximize the number":
```suggestion
 * pools until we run out of memory. Hence we limit the maximum number of
```

File: core/src/test/java/io/questdb/std/WeakObjectPoolTest.java [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        for (int i = 0; i < initSize; i++) {
+            MutablePoolElement element = tmp.get(i);
+            assertFalse(element.cleared);
+        }
+
+        pool.close();
+        assertEquals(0, pool.cache.size());
+        for (int i = 0; i < initSize; i++) {
+            MutablePoolElement element = tmp.get(i);
+            assertFalse(element.cleared);
```
Comment (puzpuzpuz) [inline core/src/test/java/io/questdb/std/WeakObjectPoolTest.java:55]: nit: would be great to add another test class that implements both `Mutable` and `Closeable` (or modify `MutablePoolElement`) and verify that it gets closed on pool close and push.

**Review by puzpuzpuz — DISMISSED** [review]

File: core/src/main/java/io/questdb/cutlass/http/HttpServer.java (hunk 1 above)
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/cutlass/http/HttpServer.java]: Many thanks!

File: core/src/test/java/io/questdb/std/WeakObjectPoolTest.java (hunk 3 above)
Comment (glasstiger) [reply core/src/test/java/io/questdb/std/WeakObjectPoolTest.java:57]: good suggestion, thank you.
just modified MutablePoolElement.

**Review by bluestreak01 — APPROVED** [review]

**Pull request merged**